pkg/gateway/services/token.go

```diff
@@ -0,0 +1,130 @@
+			Ok:     false,
+			ErrMsg: "Unable to list tokens.",
+		}, nil
+	}
+
+	var t []*pb.Token
+	for _, token := range tokens {
+		var workspaceId uint32
+		if token.WorkspaceId != nil {
+			workspaceId = uint32(*token.WorkspaceId)
+		}
+
+		updatedAt := *timestamppb.New(token.UpdatedAt)
+
+		t = append(t, &pb.Token{
+			Id:                     uint32(token.Id),
+			ExternalId:             token.ExternalId,
+			Key:                    token.Key,
+			Active:                 token.Active,
+			Reusable:               token.Reusable,
+			WorkspaceId:            &workspaceId,
+			TokenType:              token.TokenType,
+			CreatedAt:              timestamppb.New(token.CreatedAt),
+			UpdatedAt:              &updatedAt,
+			DisabledByClusterAdmin: token.DisabledByClusterAdmin,
+		})
+	}
+
+	return &pb.ListTokensResponse{
+		Tokens: t,
+		Ok:     true,
+	}, nil
+}
+
+func (gws *GatewayService) CreateToken(ctx context.Context, req *pb.CreateTokenRequest) (*pb.CreateTokenResponse, error) {
+	authInfo, _ := auth.AuthInfoFromContext(ctx)
+
+	token, err := gws.backendRepo.CreateToken(ctx, authInfo.Workspace.Id, types.TokenTypeWorkspace, true)
+	if err != nil {
+		return &pb.CreateTokenResponse{
+			Token:  &pb.Token{},
+			Ok:     false,
+			ErrMsg: "Unable to create token.",
+		}, nil
+	}
+
+	updatedAt := *timestamppb.New(token.UpdatedAt)
+	workspaceId := uint32(authInfo.Workspace.Id)
+
+	return &pb.CreateTokenResponse{
+		Token: &pb.Token{
+			Id:                     uint32(token.Id),
+			ExternalId:             token.ExternalId,
+			Key:                    token.Key,
+			Active:                 token.Active,
+			Reusable:               token.Reusable,
+			WorkspaceId:            &workspaceId,
+			TokenType:              token.TokenType,
+			CreatedAt:              timestamppb.New(token.CreatedAt),
+			UpdatedAt:              &updatedAt,
+			DisabledByClusterAdmin: token.DisabledByClusterAdmin,
+		},
+		Ok: true,
+	}, nil
+}
+
+func (gws *GatewayService) ToggleToken(ctx context.Context, req *pb.ToggleTokenRequest) (*pb.ToggleTokenResponse, error) {
+	authInfo, _ := auth.AuthInfoFromContext(ctx)
+	token, err := gws.backendRepo.ToggleToken(ctx, authInfo.Workspace.Id, req.ExternalId)
+	if err != nil {
+		return &pb.ToggleTokenResponse{
+			Token:  &pb.Token{},
+			Ok:     false,
+			ErrMsg: "Unable to toggle token.",
+		}, nil
+	}
+
+	updatedAt := *timestamppb.New(token.UpdatedAt)
+	workspaceId := uint32(authInfo.Workspace.Id)
+
+	return &pb.ToggleTokenResponse{
+		Token: &pb.Token{
+			Id:                     uint32(token.Id),
+			ExternalId:             token.ExternalId,
+			Key:                    token.Key,
+			Active:                 token.Active,
+			Reusable:               token.Reusable,
+			WorkspaceId:            &workspaceId,
+			TokenType:              token.TokenType,
+			CreatedAt:              timestamppb.New(token.CreatedAt),
+			UpdatedAt:              &updatedAt,
+			DisabledByClusterAdmin: token.DisabledByClusterAdmin,
```

Comment:
I don't know if we want this or not. @luke-lombardi what do you think?

Reply:
I removed it, but can always add it back

Reply:
Sounds good.